sdks/go/pkg/beam/core/runtime/exec/datasource.go

```diff
@@ -453,7 +454,7 @@ func (n *DataSource) Checkpoint() (SplitResult, time.Duration, bool, error) {
 // sent to this DataSource, and is used to be able to perform accurate splits
 // even if the DataSource has not yet received all its elements. A bufSize of
 // 0 or less indicates that its unknown, and so uses the current known size.
-func (n *DataSource) Split(splits []int64, frac float64, bufSize int64) (SplitResult, error) {
+func (n *DataSource) Split(ctx context.Context, splits []int64, frac float64, bufSize int64) (SplitResult, error) {
 	if n == nil {
 		return SplitResult{}, fmt.Errorf("failed to split at requested splits: {%v}, DataSource not initialized", splits)
```

Comment:
I don't think this would corrupt the split so I would return `unsuccessful split` here ditto for line 482 where `su == nil` unless both of these are programmatic errors where they should never be `nil` but might just happen to be so.

Effectively anything that is recoverable would make sense to return unsuccessful split while anything unrecoverable (e.g. you mutated the read limit for the gRPC read operation or mutated the restriction of an SDF) should return an error so that the bundle is failed.

Reply:
These are programmer errors. Unless you're going to now tell me that it's allowed for a ProcessBundleRequest to not have a root transform somehow that has a datasource.  (Speaking of which, have we added the datasource & datasink urns to the protos yet?)

Reply:
No, `beam:runner:source:v1` is always at the root. It is not part of the proto URNs list yet.

Reply:
Sounds good that these are programmatic errors.